src/TimingData.cpp

```diff
@@ -83,6 +83,265 @@ TimingData::~TimingData()
+
+				next_line.set_for_next();
+			}
+		}
+		else
+		{
+			if(event_row > 0)
+			{
+				next_line.end_beat= NoteRowToBeat(event_row);
+				float beats= next_line.end_beat - next_line.start_beat;
+				float seconds= beats * spb;
+				next_line.end_second= next_line.start_second + seconds;
+				next_line.end_expand_second= next_line.start_expand_second + seconds;
+				next_line.time_segment= curr_bpm_segment;
+				m_line_segments.push_back(next_line);
+
+				next_line.set_for_next();
+			}
+		}
+		switch(event_type)
+		{
+			case FOUND_WARP_DESTINATION:
+				// Already handled in the is_warping condition above.
+				status.is_warping= false;
+				curr_warp_segment= nullptr;
+				break;
+			case FOUND_BPM_CHANGE:
+				curr_bpm_segment= bpms[status.bpm];
+				next_line.bps= ToBPM(curr_bpm_segment)->GetBPS();
+				spb= 1.f / next_line.bps;
+				++status.bpm;
+				break;
+			case FOUND_DELAY:
+			case FOUND_STOP_DELAY:
+				next_line.end_beat= next_line.start_beat;
+				next_line.end_second= next_line.start_second +
+					ToDelay(delays[status.delay])->GetPause();
+				next_line.end_expand_second= next_line.start_expand_second;
+				next_line.time_segment= delays[status.delay];
+				m_line_segments.push_back(next_line);
+
+				next_line.set_for_next();
+				++status.delay;
+				if(event_type == FOUND_DELAY)
+				{
+					break;
+				}
+			case FOUND_STOP:
+				next_line.end_beat= next_line.start_beat;
+				next_line.end_second= next_line.start_second +
+					ToStop(stops[status.stop])->GetPause();
+				next_line.end_expand_second= next_line.start_expand_second;
+				next_line.time_segment= stops[status.stop];
+				m_line_segments.push_back(next_line);
+
+				next_line.set_for_next();
+				++status.stop;
+				break;
+			case FOUND_WARP:
+				{
+					status.is_warping= true;
+					curr_warp_segment= warps[status.warp];
+					WarpSegment* ws= ToWarp(warps[status.warp]);
+					float warp_dest= ws->GetLength() + ws->GetBeat();
+					if(warp_dest > status.warp_destination)
+					{
+						status.warp_destination= warp_dest;
+					}
+					++status.warp;
+				}
+				break;
+			default:
+				break;
+		}
+		status.last_row= event_row;
+	}
+	// m_segments_by_beat and m_segments_by_second cannot be built in the
+	// traversal above that builds m_line_segments because the vector
+	// reallocates as it grows. -Kyz
+	vector<LineSegment*>* curr_segments_by_beat= &(m_segments_by_beat[0.f]);
+	vector<LineSegment*>* curr_segments_by_second= &(m_segments_by_second[-m_fBeat0OffsetInSeconds]);
+	float curr_beat= 0.f;
+	float curr_second= 0.f;
+	for(size_t seg= 0; seg < m_line_segments.size(); ++seg)
+	{
+#define ADD_SEG(bors) \
+		if(m_line_segments[seg].start_##bors > curr_##bors) \
+		{ \
+			curr_segments_by_##bors= &(m_segments_by_##bors[m_line_segments[seg].start_##bors]); \
+			curr_##bors= m_line_segments[seg].start_##bors; \
+		} \
+		curr_segments_by_##bors->push_back(&m_line_segments[seg]);
+
+		ADD_SEG(beat);
+		ADD_SEG(second);
+#undef ADD_SEG
```

Comment:
I wonder if this macro can be replaced with a lambda function. We're now at a point where we can use such code.